src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs

```diff
@@ -199,44 +283,92 @@ public void PacketReset(SNIHandle handle, bool write, SNIPacket packet)
         /// <param name="async">Asynchronous connection</param>
         /// <param name="parallel">Attempt parallel connects</param>
         /// <returns>SNI handle</returns>
-        public SNIHandle CreateConnectionHandle(object callbackObject, string fullServerName, bool ignoreSniOpenTimeout, long timerExpire, out byte[] instanceName, byte[] spnBuffer, bool flushCache, bool async, bool parallel)
+        public SNIHandle CreateConnectionHandle(object callbackObject, string fullServerName, bool ignoreSniOpenTimeout, long timerExpire, out byte[] instanceName, ref byte[] spnBuffer, bool flushCache, bool async, bool parallel, bool isIntegratedSecurity)
         {
             instanceName = new byte[1];
             instanceName[0] = 0;
 
-            string[] serverNameParts = fullServerName.Split(':');
-
-            if (serverNameParts.Length > 2)
+            SNIHandle sniHandle = null;
+            if (!fullServerName.Contains(":"))
             {
-                SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.INVALID_PROV, 0, SNICommon.InvalidConnStringError, string.Empty);
-                return null;
+                // default to using tcp if no protocol is provided
+                sniHandle = CreateTcpHandle(fullServerName, timerExpire, callbackObject, parallel, ref spnBuffer, isIntegratedSecurity);
             }
-
-            // Default to using tcp if no protocol is provided
-            if (serverNameParts.Length == 1)
-            {
-                return CreateTcpHandle(serverNameParts[0], timerExpire, callbackObject, parallel);
-            }
-
-            switch (serverNameParts[0])
+            else
             {
-                case TdsEnums.TCP:
-                    return CreateTcpHandle(serverNameParts[1], timerExpire, callbackObject, parallel);
-
-                case TdsEnums.NP:
-                    return CreateNpHandle(serverNameParts[1], timerExpire, callbackObject, parallel);
+                // when tcp protocol is specified
+                if (fullServerName.ToLower().StartsWith(TdsEnums.TCP + ":"))
```

Comment:
There are too many control statements around connection based on the result of string parsing.
Consider refactoring this server string parsing into another method which returns relevant information like protocol, hostname, port name and instance name. 
Based on this structured information, the connection decision should be taken.